tests/tests/detail/not_tuple.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 Bronek Kozicki
+//
+// Distributed under the ISC License. See accompanying file LICENSE.md
+// or copy at https://opensource.org/licenses/ISC
+
+#include "functional/detail/not_tuple.hpp"
+
+#include <catch2/catch_all.hpp>
+
+namespace fn::detail {
+
+struct Foo final {};
+static_assert([](auto &&v) constexpr -> bool {
+  return requires { get<0>(v); } //
+         && !requires { get<1>(v); };
+}(not_tuple<1, Foo>{}));
+
+static_assert([](auto &&v) constexpr -> bool {
+  return requires { get<0>(v); }    //
+         && requires { get<1>(v); } //
+         && !requires { get<2>(v); };
+  ;
+}(not_tuple<2, Foo, Foo>{}));
+
+static_assert([](auto &&v) constexpr -> bool {
+  return requires { get<0>(v); }    //
+         && requires { get<1>(v); } //
+         && requires { get<2>(v); } //
+         && !requires { get<3>(v); };
+  ;
+}(not_tuple<3, Foo, Foo, Foo>{}));
+
+static_assert([](auto &&v) constexpr -> bool {
+  return requires { get<0>(v); }    //
+         && requires { get<1>(v); } //
+         && requires { get<2>(v); } //
+         && requires { get<3>(v); } //
+         && !requires { get<4>(v); };
+  ;
+}(not_tuple<4, Foo, Foo, Foo, Foo>{}));
+
+// clang-format off
+struct A { int v; };
+static_assert(std::is_same_v<decltype(get<0>(std::declval<not_tuple<1, A>>())), A &&>);
```

Comment:
Would we need `volatile` and all variations here too or does `not_tuple` only support the ones that are actually used by the `fn` library deliberately?

Reply:
same as [above](https://eel.is/c++draft/depr.volatile.type)